larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+    LArHitWidthHelper::ClusterToParametersMap* pClusterToParametersMap = pClusterToParametersMapStore->GetMap();
+
+    if(pClusterToParametersMap->empty())
+        throw StatusCodeException(STATUS_CODE_NOT_INITIALIZED);
+
+    // ATTN this method assumes that clusters have been sorted by extremal x position (low x -> high x)
+    for (ClusterVector::const_iterator iterCurrentCluster = clusterVector.begin(); iterCurrentCluster != clusterVector.end(); ++iterCurrentCluster)
+    {
+        const Cluster *const pCurrentCluster = *iterCurrentCluster;
+        const LArHitWidthHelper::ClusterToParametersMap::const_iterator currentParametersIter = pClusterToParametersMap->find(pCurrentCluster);
+
+        if(currentParametersIter == pClusterToParametersMap->end())
+            throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+        for (ClusterVector::const_iterator iterTestCluster = iterCurrentCluster; iterTestCluster != clusterVector.end(); ++iterTestCluster)
+        {
+            if (iterCurrentCluster == iterTestCluster)
+                continue;
+
+	    const Cluster *const pTestCluster = *iterTestCluster;
+            const LArHitWidthHelper::ClusterToParametersMap::const_iterator testParametersIter = pClusterToParametersMap->find(pTestCluster);
+
+	    if(testParametersIter == pClusterToParametersMap->end())
+                throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+            if (!this->AreClustersAssociated(currentParametersIter->second, testParametersIter->second))
+	        continue;
+
+            clusterAssociationMap[pCurrentCluster].m_forwardAssociations.insert(pTestCluster);
+            clusterAssociationMap[pTestCluster].m_backwardAssociations.insert(pCurrentCluster);
+        }
+    }
+
+    // remove all 'shortcut' routes
+    this->CleanupClusterAssociations(clusterVector, clusterAssociationMap);
+
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void HitWidthClusterMergingAlgorithm::CleanupClusterAssociations(const ClusterVector &clusterVector, ClusterAssociationMap &clusterAssociationMap) const
+{
+
+  // Create temporary map so can delete elements whilst still iterating over them
+  ClusterAssociationMap tempMap(clusterAssociationMap);
+
+  for(const Cluster *const pCluster : clusterVector) 
+  {
+      const ClusterAssociationMap::const_iterator primaryMapIter = clusterAssociationMap.find(pCluster);
+
+      if(primaryMapIter == clusterAssociationMap.end())
+          continue;
+
+      const ClusterSet &primaryForwardAssociations(primaryMapIter->second.m_forwardAssociations);
+
+      // loop through the primary associations of each cluster
+      // remove clusters that are present in secondary associations of other primary clusters 
+      for(const Cluster *const pConsideredCluster : primaryForwardAssociations)
+      {
```

Comment:
Unfortunately, and this is a weakness of this algorithm, the use of std::unordered_set means that `primaryForwardAssociations` isn't guaranteed to be the same each time you run. So, you must copy from the ClusterSet into a ClusterVector or ClusterList and call one of the sort functions in LArClusterHelper, which have been put together carefully to handle ties in many quantities and guarantee a defined order. You can then loop over the ClusterVector/List and look up what's in the set (or any other maps) if needed. Sets and maps are great for lookup, and that's how we should use them. Any time we loop over either, we need to think about whether we should.

Reply:
I have made these changes. Just for learning purposes - In this loop the order of looping doesn't matter so long as every element is considered. Will this not happen in the unordered_set, is it unordered in the sense that the ordering can change during the looping process?

Reply:
Unfortunately it doesn't always turn out to be so simple, because you can get e.g. floating point rounding in calculations that can mean that the result can differ if the calculations are performed in different orders (only by a small amount, but that might be enough to trigger a different macroscopic decision). The only way to guarantee it's completely reproducible is to walk over the elements in the same order. It also makes things much easier to debug in future, and is a universally preferred. Any time we make a loop, we should think about whether we're looping over objects in a defined and reproducible order (this can be just the order in which things are passed in, but so long as they're passed in in the same order on every machine, it's fine).